src/ghost_borrow_mut.rs

```diff
@@ -255,23 +311,113 @@ generate_public_instance!(a, b, c, d, e, f, g, h, i, j, k, l ; T0, T1, T2, T3, T
+    #[inline(always)]
+    fn split<'a>(spans: &[sealed::Span], singles: &'a mut [*const u8], slices: &'a mut [(*const u8, *const u8)])
+        -> (&'a mut [*const u8], &'a mut [(*const u8, *const u8)])
+    {
+        let mut single_index = 0;
+        let mut slice_index = 0;
+
+        for span in spans {
+            use sealed::Span::*;
+
+            match *span {
+                Single(single) => {
+                    singles[single_index] = single;
+                    single_index += 1;
+                }
+                Slice(start, end) => {
+                    slices[slice_index] = (start, end);
+                    slice_index += 1;
                 }
             }
         }
-    } else {
-        arr.sort_unstable();
-        for i in 0..(N - 1) {
-            if core::ptr::eq(arr[i] as *const (), arr[i + 1] as *const ()) {
+
+        let singles = &mut singles[..single_index];
+        let slices = &mut slices[..slice_index];
+
+        (singles, slices)
+    }
+
+    //  Pre-condition: `singles` is sorted.
+    fn check_singles(singles: &[*const u8]) -> Result<(), GhostAliasingError> {
+        for window in singles.windows(2) {
+            if window[0] == window[1] {
                 return Err(GhostAliasingError);
             }
         }
+
+        Ok(())
+    }
+
+    //  Pre-condition: `slices` is sorted.
+    fn check_slices(slices: &[(*const u8, *const u8)]) -> Result<(), GhostAliasingError> {
+        fn overlap((a, b): (*const u8, *const u8), (c, d): (*const u8, *const u8)) -> bool {
+            //  Ranges [A, B) and [C, D) overlap if:
+            //
+            //  -   [A, B) is a super-range of [C, D), ie A <= C && D <= B.
+            //  -   [C, D) is a super-range of [A, B), ie C <= A && B <= D.
+            //  -   C belongs to [A, B), ie C >= A && C < B.
+            //  -   D belongs to [A, B), ie D >= A && D < B.
+            //
+            //  That's a lot of conditions, so checking the absence of overlapping is surprisingly easier. The ranges do
+            //  not overlap if B <= C or A >= D.
+            //
+            //  Hence they do overlap if !(B <= C or A >= D) <=> !(B <= C) and !(A >= D) <=> B > C and A < D.
+
+            b > c && a < d
+        }
+
+        for window in slices.windows(2) {
+            if overlap(window[0], window[1]) {
+                return Err(GhostAliasingError);
+            }
+        }
+
+        Ok(())
+    }
+
+    //  Pre-condition: `singles` and `slices` are sorted.
+    fn check_singles_in_slices(singles: &[*const u8], mut slices: &[(*const u8, *const u8)]) -> Result<(), GhostAliasingError> {
+        for &single in singles {
+            //  Find the first element of `slices` whose end-of-range is strictly after `single`.
+            let Some(start) = slices.iter().position(|slice| single < slice.1) else { return Ok(()) };
+
+            slices = &slices[start..];
+
+            for slice in slices {
+                if single >= slice.1 { break }
```

Comment:
This never happens. In fact, this function has quadratic complexity. This should have been `single < slice.0`, and the whole loop should have been an `if`.